silence/silence.go

```diff
@@ -0,0 +1,792 @@
+		return errNotSupported
+	}
+}
+
+// QMatches returns silences that match the given label set.
+func QMatches(set map[string]string) QueryParam {
+	return func(q *query) error {
+		f := func(s *pb.Silence, _ *timestamp.Timestamp) (bool, error) {
+			// TODO(fabxc): we compile every regexp matcher of a silence
+			// each time we check it against a label set.
+			// This could be notably slower than the old caching behavior.
+			// With the protobuf type not being extensible, we need a more
+			// efficient solution without wrapping the silence in another layer.
+			for _, m := range s.Matchers {
+				switch m.Type {
+				case pb.Matcher_EQUAL:
+					if set[m.Name] != m.Pattern {
+						return false, nil
+					}
+				case pb.Matcher_REGEXP:
+					re, err := regexp.Compile(m.Pattern)
+					if err != nil {
+						return false, err
+					}
+					if !re.MatchString(set[m.Name]) {
+						return false, nil
+					}
+				}
+			}
+			// All matchers applied to the given set and the silence
+			// passes as a result.
+			return true, nil
+		}
+		q.filters = append(q.filters, f)
+		return nil
+	}
+}
+
+// SilenceState describes the state of a silence based on its time range.
+type SilenceState string
+
+// The only possible states of a silence w.r.t a timestamp.
+const (
+	StateActive  SilenceState = "active"
+	StatePending              = "pending"
+	StateExpired              = "expired"
+)
+
+// getState returns a silence's SilenceState at the given timestamp.
+func getState(sil *pb.Silence, ts *timestamp.Timestamp) SilenceState {
+	if protoBefore(ts, sil.StartsAt) {
+		return StatePending
+	}
+	if protoBefore(sil.EndsAt, ts) {
+		return StateExpired
+	}
+	return StateActive
+}
+
+// QState filters queried silences by the given states.
+func QState(states ...SilenceState) QueryParam {
+	return func(q *query) error {
+		f := func(sil *pb.Silence, now *timestamp.Timestamp) (bool, error) {
+			s := getState(sil, now)
+			for _, ps := range states {
+				if s == ps {
+					return false, nil
+				}
+			}
+			return true, nil
+		}
+		q.filters = append(q.filters, f)
+		return nil
+	}
+}
+
+// Query for silences based on the given query parameters.
+func (s *Silences) Query(params ...QueryParam) ([]*pb.Silence, error) {
+	q := &query{}
+	for _, p := range params {
+		if err := p(q); err != nil {
+			return nil, err
+		}
+	}
+	nowpb, err := s.nowProto()
+	if err != nil {
+		return nil, err
+	}
+	return s.query(q, nowpb)
+}
+
+func (s *Silences) query(q *query, now *timestamp.Timestamp) ([]*pb.Silence, error) {
+	// If we have an ID constraint, all silences are our base set.
+	// This and the use of post-filter functions is the
+	// the trivial solution for now.
+	var res []*pb.Silence
+
+	s.mtx.RLock()
+	if q.ids != nil {
+		for _, id := range q.ids {
+			if s, ok := s.st[string(id)]; ok {
+				res = append(res, s.Silence)
+			}
+		}
+	} else {
+		for _, sil := range s.st {
+			res = append(res, sil.Silence)
+		}
+	}
+	s.mtx.RUnlock()
```

Comment:
We never modify existing silence objects (at least that was the intent, anything else is a bug). What's protected is the map access. So once we've created our silence slice, reading anything in these silences is fine.

We cannot get a memory race here – but a semantical one as we could expire a silence (by ID) while we are considering its shallow copy below. Going by that, we can also run into issues where a silence expires between checking different alerts.
Given the timing precision we are generally concerned with, this seems fine. The alternative for now would be to lock up everything for the entire time, by adding transactions or batching. That doesn't seem feasible in general.

Time is everywhere in Alertmanager and with all the artefacts this comes with – especially in the eventually consistent HA setting. We might even get a silence from a different AM that started in the past where we have notified about affected alerts already.